### ISSUE TITLE
Balrog textbox oddity

### ISSUE THREAD

**Issue by Fiefio-real** [doukutsu-rs/doukutsu-rs#278]
After the first fight with Balrog the congratulatory message is displayed in a textbox. While that textbox is still on screen the game spawns 5 xp triangles. In Doukutsu-rs each time an xp triangle is spawned you can hear the text writing sound effect and the textbox message gets scrolled up. There's also a longer delay between each xp spawn.
Cave Story:
[balrogorg.webm](https://github.com/user-attachments/assets/643a84a2-2e13-4501-bf0b-ae012428d033)
Doukutsu-rs:
[balrog.webm](https://github.com/user-attachments/assets/e23edee1-c232-4dc1-b527-de2ae632cf33)

I think there was a similar thing happening right before Balrog's fight in the Boulder Chamber, I'll try to get some footage when I have more time.

**Comment by periwinkle9** [comment]
This is due to (what is arguably) a bug in the original freeware version that causes extra script to be parsed when a textbox scroll is meant to happen, which can sometimes interfere with the textbox scrolling. Whether we really want to replicate this behavior in d-rs or not depends on how much we care about accuracy to the original (in spite of bugs) vs. "correct"/intended behavior...

For context, here's the event in question:
```
#0400
<KEY<FL-0361<FL-0362
<SMC<CNP0400:0012:0000<ANP0400:0020:0004<WAI0140
<PRI<MSG<FAC0005Just you wait!<NOD<FAC0000<KEY<CLO
<WAI0100<CMU0015
<MSG
Fought off Balrog!<WAI0140<NOD
<SOU0045<CNP0150:0001:0000<WAI0004
<SOU0045<CNP0151:0001:0000<WAI0004
<SOU0045<CNP0152:0001:0000<WAI0004
<SOU0045<CNP0153:0001:0000<WAI0004
<SOU0045<CNP0154:0001:0000
<CMU0008<END
```
The last few lines (with the `<CNP`s) are what spawn the xp crystals. Notice how they are on separate lines, but the textbox above it has not been closed with `<CLO`. Thus, this technically _should_ result in newlines being written to the textbox, which is what you see in d-rs. (The extra delay between the xp crystals spawning is due to the additional time it takes to scroll the textbox.) However, in freeware, the bug causes the line after where the textbox should start scrolling to be parsed early, and ultimately the `<WAI0004` overrides the mode change and ends up canceling the textbox scroll altogether.

To demonstrate further, suppose we add some text in the middle of this:
```
<MSG
Fought off Balrog!<WAI0140<NOD
<SOU0045<CNP0150:0001:0000<WAI0004
<SOU0045<CNP0151:0001:0000<WAI0004
<SOU0045<CNP0152:0001:0000<WAI0004
lol good game Pixel<SOU0045<CNP0153:0001:0000<WAI0004
<SOU0045<CNP0154:0001:0000
<CMU0008<END
```
Just looking at the script, you would think that this should print "lol good game Pixel" some number of lines below "Fought off Balrog!". (I assume this is the behavior you would see in d-rs, but I haven't tested it.)
However, if you run this script in freeware, this is what you'll actually see:
![textbug](https://github.com/user-attachments/assets/639b8dfc-fccb-41ed-ad10-c8df4064f762)

**Comment by Fiefio-real** [comment]
Here's the other instance of this phenomenon I was talking about:
[balrog2.webm](https://github.com/user-attachments/assets/c49a72ba-70f8-4a5a-86bf-9438b851cef8)
In Cave Story textbox closing, Balrog changing his sprite and music starting playing all happen at the same time. In this case the music starts playing and Balrog switches his sprite while the textbox is still on screen and scrolls the text.

**Comment by periwinkle9** [comment]
Yup, that's another instance of the same phenomenon that I explained.
```
One, two...<NOD<ANP0310:0080:0000<MYB0002<ANP0300:0030:0000<SOU0072<CLR<FAC0023ARE YOU EVEN LISTENING
TO ME?!<NOD
<ANP0310:0000:0000<CMU0004<CNP0310:0169:0000<BSL0310
<FL+0762<FL+0763
<END
```
In this case, there are 3 newlines between the last line of text and the `<END`, which d-rs prints properly, resulting in the textbox scrolling twice before the cutscene ends. Vanilla Cave Story also recognizes that the textbox should scroll, but instead of pausing immediately to do so, it continues parsing the event, until ultimately reaching the `<END` that kills the whole event before any of that is able to occur.
